byzcoin/heartbeat.go

```diff
@@ -70,49 +71,71 @@ func (r *heartbeats) exists(key string) bool {
 	return ok
 }
 
+// updateTimeout stores the new timeout and resets the timer.
+func (r *heartbeats) updateTimeout(key string, timeout time.Duration) {
+	r.Lock()
+	defer r.Unlock()
+	h, ok := r.heartbeatMap[key]
+	if !ok {
+		return
+	}
+	if h.timeout != timeout {
+		h.updateTO <- timeout
+	}
+}
+
 func (r *heartbeats) start(key string, timeout time.Duration, timeoutChan chan string) error {
 	r.Lock()
 	defer r.Unlock()
 	if _, ok := r.heartbeatMap[key]; ok {
 		return errors.New("key already exists")
 	}
 
-	beatChan := make(chan bool)
-	closeChan := make(chan bool, 1)
-	getTimeChan := make(chan chan time.Time, 1)
+	r.heartbeatMap[key] = &heartbeat{
+		beatChan:    make(chan bool),
+		closeChan:   make(chan bool, 1),
+		getTimeChan: make(chan chan time.Time, 1),
+		timeout:     timeout,
+		timeoutChan: timeoutChan,
+		updateTO:    make(chan time.Duration),
+	}
 
 	r.wg.Add(1)
-	go func() {
+	go func(h *heartbeat) {
 		defer r.wg.Done()
 		currTime := time.Now()
-		to := time.After(timeout)
+		to := time.NewTimer(h.timeout)
 		for {
 			select {
```

Comment:
This is tricky, how about a small blcok comment here to explain the interaction between to.Stop() and t.C for the next guy. :)